sdk/python/kfp/v2/compiler/compiler.py

```diff
@@ -59,68 +64,533 @@ def my_pipeline(a: int = 1, b: str = "default value"):
+        return "inputs.parameters['{input_name}'].{value_field}".format(
+            input_name=input_name,
+            value_field=type_utils.get_parameter_type_field_name(
+                value_or_reference.param_type))
+      else:
+        raise NotImplementedError(
+            'Use artifact as dsl.Condition operand is not implemented yet.')
+    else:
+      if isinstance(value_or_reference, str):
+        return "'{}'".format(value_or_reference)
+      else:
+        return str(value_or_reference)
+
+  def _group_to_dag_template(
+      self,
+      group: dsl.OpsGroup,
+      inputs: Dict[str, List[Tuple[dsl.PipelineParam, str]]],
+      outputs: Dict[str, List[Tuple[dsl.PipelineParam, str]]],
+      dependencies: Dict[str, List[GroupOrOp]],
+      pipeline_spec: pipeline_spec_pb2.PipelineSpec,
+      deployment_config: pipeline_spec_pb2.PipelineDeploymentConfig,
+      root_name: str,
+  ) -> None:
+    """Generate template given an OpsGroup.
+
+    inputs, outputs, dependencies are all helper dicts.
+    """
+    group_component_name = dsl_utils.sanitize_component_name(group.name)
+
+    if group.name == root_name:
+      group_component_spec = pipeline_spec.root
+    else:
+      group_component_spec = pipeline_spec.components[group_component_name]
+
+    # Generate component inputs spec.
+    if inputs.get(group.name, None):
+      dsl_component_spec.build_component_inputs_spec(
+          group_component_spec, [param for param, _ in inputs[group.name]])
+
+    # Generate component outputs spec.
+    if outputs.get(group.name, None):
+      group_component_spec.output_definitions.CopyFrom(
+          dsl_component_spec.build_component_outputs_spec(
+              [param for param, _ in outputs[group.name]]))
+
+    # Generate task specs and component specs for the dag.
+    sub_groups = group.groups + group.ops
+    for sub_group in sub_groups:
+      subgroup_task_spec = getattr(sub_group, 'task_spec',
+                                   pipeline_spec_pb2.PipelineTaskSpec())
+      subgroup_component_spec = getattr(sub_group, 'component_spec',
+                                        pipeline_spec_pb2.ComponentSpec())
+      is_recursive_subgroup = (
+          isinstance(sub_group, dsl.OpsGroup) and sub_group.recursive_ref)
+      # Special handling for recursive subgroup: use the existing opsgroup name
+      if is_recursive_subgroup:
+        sub_group_key = sub_group.recursive_ref.name
+      else:
+        sub_group_key = sub_group.name
+
+      subgroup_task_spec.task_info.name = dsl_utils.sanitize_task_name(
+          sub_group_key)
+      # human_name exists for ops only, and is used to de-dupe component spec.
+      subgroup_component_name = dsl_utils.sanitize_component_name(
+          getattr(sub_group, 'human_name', sub_group_key))
+      subgroup_task_spec.component_ref.name = subgroup_component_name
+
+      if isinstance(sub_group, dsl.OpsGroup) and sub_group.type == 'condition':
```

Comment:
Can the logic below be split into helper functions to help with readability?

Reply:
Extracted the task_spec inputs construction part, which is similar to component_spec input_definitions construction.